content/master/concepts/composition-functions.md

```diff
@@ -1,1026 +1,699 @@
-
- * Be packaged as an OCI image, where the `ENTRYPOINT` is the Function.
- * Accept input in the form of a `FunctionIO` document on stdin.
- * Return the `FunctionIO` they were passed, optionally mutated, on stdout.
- * Run within the constraints specified by the Composition that includes them,
-   such as timeouts, compute, network access.
 
-This means Functions may be written using a general-purpose programming language
-like Python, Go, or TypeScript. They may also be implemented using a shell
-script, or an existing tool like Helm or Kustomize.
+The `functions.yaml` file contains the Functions the Composition references in
+its pipeline steps:
 
-### FunctionIO
-
-When a Composition Function runner like `xfn` runs your Function it will write
-`FunctionIO` to its stdin. A `FunctionIO` is a Kubernetes style YAML manifest.
-It's not a custom resource (it never gets created in the API server) but it
-follows Kubernetes conventions.
-
-A `FunctionIO` consists of:
-
-* An optional, arbitrary `config` object.
-* The `observed` state of the XR, any existing composed resources, and their connection details.
-* The `desired` state of the XR and any composed resources.
-* Optional `results` of the Function pipeline.
+```yaml
+---
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-patch-and-transform
+spec:
+  package: xpkg.upbound.io/crossplane-contrib/function-patch-and-transform:v0.1.4
+```
+{{</expand>}}
+
+The Crossplane CLI uses Docker Engine to run functions. You can change how the
+Crossplane CLI run a function by adding an annotation in `functions.yaml`. Add
+the `render.crossplane.io/runtime` annotation to a Function to change how it's
+run.
+
+`crossplane beta render` supports two `render.crossplane.io/runtime` values:
+
+* `Docker` (the default) connects to Docker Engine. It uses Docker to pull and
+  run function runtimes.
+* `Development` connects to a function runtime you have run manually.
+
+When you use the {{<hover label="development" line="6">}}Development{{</hover>}}
+runtime the Crossplane CLI ignores the Function's {{<hover label="development"
+line="8">}}package{{</hover>}}. Instead it expects you to make sure the function
+is listening on localhost port 9443. The function must be listening without gRPC
+transport security. Most function SDKs let you run a function with the
+`--insecure` flag to disable transport security. For example you can run a Go
+function locally using `go run . --insecure`.
+
+```yaml {label="development"}
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-patch-and-transform
+  annotation:
+    render.crossplane.io/runtime: Development
+spec:
+  package: xpkg.upbound.io/crossplane-contrib/function-patch-and-transform:v0.1.4
+```
 
-Here's a brief example of a `FunctionIO`:
+{{<hint "tip">}}
+Use the `Development` runtime when you
+[write a composition function](#write-a-composition-function) to test your
+function end-to-end.
```

Comment:
Can you add `end-to-end` to 
`utils/vale/styles/Crossplane/spelling-exceptions.txt`

to fix the vale error here?